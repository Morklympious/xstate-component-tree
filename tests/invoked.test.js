"use strict";

const { Machine : createMachine, interpret } = require("xstate");
const trees = require("./util/trees.js");
const component = require("./util/component.js");

describe("xstate-component-tree", () => {
    describe("invoked child machines", () => {
        it("should be supported", async () => {
            const childMachine = createMachine({
                initial : "child",

                states : {
                    child : {
                        meta : {
                            component : component("child"),
                        },
                    },
                },
            });

            const testMachine = createMachine({
                initial : "one",

                states : {
                    one : {
                        invoke : {
                            id  : "child",
                            src : childMachine,
                        },

                        meta : {
                            component : component("one"),
                        },
                    },
                },
            });

            const service = interpret(testMachine);

            const states = trees(service);
            
            await states(2);
        });

        it("should rebuild on transitions", async () => {
            const childMachine = createMachine({
                initial : "child1",

                states : {
                    child1 : {
                        meta : {
                            component : component("child1"),
                        },

                        on : {
                            NEXT : "child2",
                        },
                    },

                    child2 : {
                        meta : {
                            component : component("child2"),
                        },
                    },
                },
            });

            const testMachine = createMachine({
                initial : "one",

                states : {
                    one : {
                        invoke : {
                            id  : "child",
                            src : childMachine,
                            
                            autoForward : true,
                        },

                        meta : {
                            component : component("one"),
                        },
                    },
                },
            });

            const service = interpret(testMachine);

            const states = trees(service);

            await states(2);

            service.send("NEXT");

            await states();
        });

        it("should support nested invoked machines", async () => {
            const grandchildMachine = createMachine({
                initial : "grandchild",

                states : {
                    grandchild : {
                        meta : {
                            component : component("grandchild"),
                        },
                    },
                },
            });

            const childMachine = createMachine({
                initial : "child",

                states : {
                    child : {
                        invoke : {
                            id  : "grandchild",
                            src : grandchildMachine,

                            autoForward : true,
                        },

                        meta : {
                            component : component("child"),
                        },
                    },
                },
            });

            const testMachine = createMachine({
                initial : "one",

                states : {
                    one : {
                        invoke : {
                            id  : "child",
                            src : childMachine,
                            
                            autoForward : true,
                        },

                        meta : {
                            component : component("one"),
                        },
                    },
                },
            });

            const service = interpret(testMachine);

            const states = trees(service);

            await states(3);
        });

        it("should rebuild on nested invoked machine transitions", async () => {
            const grandchildMachine = createMachine({
                initial : "grandchild1",

                states : {
                    grandchild1 : {
                        meta : {
                            component : component("grandchild1"),
                        },

                        on : {
                            NEXT : "grandchild2",
                        },
                    },

                    grandchild2 : {
                        meta : {
                            component : component("grandchild2"),
                        },
                    },
                },
            });

            const childMachine = createMachine({
                initial : "child",

                states : {
                    child : {
                        invoke : {
                            id  : "grandchild",
                            src : grandchildMachine,

                            autoForward : true,
                        },

                        meta : {
                            component : component("child"),
                        },
                    },
                },
            });

            const testMachine = createMachine({
                initial : "one",

                states : {
                    one : {
                        invoke : {
                            id  : "child",
                            src : childMachine,
                            
                            autoForward : true,
                        },

                        meta : {
                            component : component("one"),
                        },
                    },
                },
            });

            const service = interpret(testMachine);

            const states = trees(service);

            await states(3);

            service.send("NEXT");

            await states();
        });
    });
});
