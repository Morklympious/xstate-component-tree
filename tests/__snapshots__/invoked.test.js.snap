// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xstate-component-tree invoked child machines should be supported 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree invoked child machines should be supported 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
]
`;

exports[`xstate-component-tree invoked child machines should rebuild on nested invoked machine transitions 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree invoked child machines should rebuild on nested invoked machine transitions 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
]
`;

exports[`xstate-component-tree invoked child machines should rebuild on nested invoked machine transitions 3`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "grandchild1",
      },
    ],
    "id": "grandchild",
  },
]
`;

exports[`xstate-component-tree invoked child machines should rebuild on nested invoked machine transitions 4`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "grandchild2",
      },
    ],
    "id": "grandchild",
  },
]
`;

exports[`xstate-component-tree invoked child machines should rebuild on transitions 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree invoked child machines should rebuild on transitions 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child1",
      },
    ],
    "id": "child",
  },
]
`;

exports[`xstate-component-tree invoked child machines should rebuild on transitions 3`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child2",
      },
    ],
    "id": "child",
  },
]
`;

exports[`xstate-component-tree invoked child machines should support nested invoked machines 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree invoked child machines should support nested invoked machines 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
]
`;

exports[`xstate-component-tree invoked child machines should support nested invoked machines 3`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "grandchild",
      },
    ],
    "id": "grandchild",
  },
]
`;
